src/topology_sampler.cpp

```diff
@@ -0,0 +1,126 @@
+  session.result_.AddVertex({node.Id(), node.GetBitset()});
+  SampleRootward(session, node);
+  SampleLeafward(session, node, Clade::Left);
+  SampleLeafward(session, node, Clade::Right);
+  session.result_.ConnectAllVertices();
+  auto root = session.result_.FindRoot();
+  if (!root.has_value()) Failwith("No root found");
+  return BuildTree(session, root.value().get());
+}
+
+void TopologySampler::SetSeed(uint64_t seed) { mersenne_twister_.SetSeed(seed); }
+
+void TopologySampler::VisitNode(SamplingSession& session, SubsplitDAGNode node,
+                                Direction direction, Clade clade) {
+  session.result_.AddVertex({node.Id(), node.GetBitset()});
+  switch (direction) {
+    case Direction::Rootward:
+      SampleLeafward(session, node, Clade::Left);
+      SampleLeafward(session, node, Clade::Right);
+      break;
+    case Direction::Leafward:
+      SampleRootward(session, node);
+      SampleLeafward(session, node, Opposite(clade));
+      break;
+  }
+}
+
+void TopologySampler::SampleRootward(SamplingSession& session, SubsplitDAGNode node) {
+  auto left = node.GetLeftRootward();
+  auto right = node.GetRightRootward();
+  if (left.empty() && right.empty()) {
+    // reached root
+    return;
+  }
+  auto parent = SampleParent(session, left, right);
+  session.result_.AddLine({parent.second.GetId(), parent.second.GetParent(),
+                           parent.second.GetChild(), parent.second.GetClade()});
+  VisitNode(session, parent.first, Direction::Leafward, parent.second.GetClade());
+}
+
+void TopologySampler::SampleLeafward(SamplingSession& session, SubsplitDAGNode node,
+                                     Clade clade) {
+  auto neighbors = node.GetNeighbors(Direction::Leafward, clade);
+  if (neighbors.empty()) {
+    // reached leaf
+    return;
+  }
+  auto child = SampleChild(session, neighbors);
+  session.result_.AddLine({child.second.GetId(), child.second.GetParent(),
+                           child.second.GetChild(), child.second.GetClade()});
+  VisitNode(session, child.first, Direction::Rootward, clade);
+}
+
+std::pair<SubsplitDAGNode, ConstLineView> TopologySampler::SampleParent(
+    SamplingSession& session, ConstNeighborsView left, ConstNeighborsView right) {
+  std::vector<double> weights;
+  weights.resize(left.size() + right.size());
+  size_t i = 0;
+  for (auto parent = left.begin(); parent != left.end(); ++parent)
+    weights[i++] = session.inverted_probabilities_[parent.GetEdge()];
+  for (auto parent = right.begin(); parent != right.end(); ++parent)
+    weights[i++] = session.inverted_probabilities_[parent.GetEdge()];
+  std::discrete_distribution<> distribution(weights.begin(), weights.end());
+  i = static_cast<size_t>(distribution(mersenne_twister_.GetGenerator()));
+  if (i < left.size()) {
+    auto parent = left.begin();
+    std::advance(parent, i);
+    return {session.dag_.GetDAGNode(parent.GetNodeId()),
+            session.dag_.GetDAGEdge(parent.GetEdge())};
+  }
```

Comment:
```suggestion
  } // else
```

I like to put this around early returns